File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-1.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} val
     * @return {TreeNode}
     */
    insertIntoBST(root: TreeNode | null, val: number): TreeNode {
        let node = new TreeNode(val);
        if(root == null){
            return node;
        }
        let curr = root;
        let prev = null;
        while(curr){
            prev = curr;
            if (val > curr.val){
                curr = curr.right;
            } else {
                curr = curr.left;
            }
        }
        if(val > prev.val){
            prev.right = node;
        } else {
            prev.left = node;
        }

        return root;
    }
}

File: Data Structures & Algorithms/binary-search-tree-iterator/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class BSTIterator {
    private TreeNode cur;
    private Stack<TreeNode> stk;

    public BSTIterator(TreeNode root) {
        this.stk = new Stack<TreeNode>();
        this.cur = root;
    }
    
    public int Next() {
        while(this.cur != null){
            this.stk.Push(cur);
            this.cur = this.cur.left;
        }
        this.stk.TryPop(out this.cur);
        var ans = this.cur.val;
        this.cur = this.cur.right;
        return ans;
    }
    
    public bool HasNext() {
        return this.cur != null || this.stk.Count > 0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */